R/git.R

```diff
@@ -231,6 +238,123 @@ choose_protocol <- function() {
+#' provide the user a way to intervene:
+#'   * `git2r_credentials()` returns any `credentials` that have been registered
+#'     with `use_git2r_credentials()` and, otherwise, implements usethis's
+#'     default strategy.
+#'   * `use_git2r_credentials()` allows you to register `credentials` explicitly
+#'     for use in all usethis functions in an R session. Do this only after
+#'     proven failure of the defaults.
+#'
+#' @section Default credentials:
+#'
+#'   If the default behaviour of usethis + git2r works, rejoice and leave well
+#'   enough alone. Keep reading if you need more control or understanding.
+#'
+#' @section SSH credentials:
+#'
+#'   For `protocol = "ssh"`, by default, usethis passes `NULL` credentials
+#'   to git2r. This will work if you have the exact configuration expected by
+#'   git2r:
+#'     1. Your public and private keys are in the default locations,
+#'     `~/.ssh/id_rsa.pub` and `~/.ssh/id_rsa`, respectively.
+#'     1. All the relevant software agrees on the definition of `~/`, i.e.
+#'     your home directory. This is harder than it sounds on Windows.
+#'     1. Your `ssh-agent` is configured to manage your SSH passphrase, if you
+#'     have one. This too can be a problem on Windows.
+#'   Read more about SSH setup in [Happy
+#'   Git](http://happygitwithr.com/ssh-keys.html), especially the
+#'   [troubleshooting
+#'   section](http://happygitwithr.com/ssh-keys.html#ssh-troubleshooting).
+#'
+#'   If the `NULL` default doesn't work, you can make `credentials` explicitly
+#'   with [git2r::cred_ssh_key()] and register that with
+#'   `use_git2r_credentials()` for the rest of the session:
+#' ```
+#' my_cred <- git2r::cred_ssh_key(
+#'    publickey  = "path/to/your/id_rsa.pub",
+#'    privatekey = "path/to/your/id_rsa",
+#'    passphrase = "my_supersecret_passphrase"
+#' )
+#' use_git2r_credentials(credentials = my_cred)
+#' ```
+#'   For the remainder of the session, `git2r_credentials()` will return
+#'   `my_cred`.
+#'
+#' @section HTTPS credentials:
+#'
+#'   For `protocol = "https"`, we must send username and password. It is
+#'   possible that your OS has cached this and git2r will successfully use that.
+#'   However, usethis can offer even more chance of success in the HTTPS case.
+#'   GitHub also accepts a personal access token (PAT) via HTTPS. If
+#'   `credentials = NULL` and a PAT is available, we send it. Preference is
+#'   given to any `auth_token` that is passed explicitly. Otherwise, the
+#'   environment variables `GITHUB_PAT` and `GITHUB_TOKEN` are consulted, in
+#'   that order. If a PAT is found, we make an HTTPS credential with
+#'   [git2r::cred_user_pass()]. The PAT is sent as the password and dummy text
+#'   is sent as the username (the PAT is what really matters in this case). You
+#'   can also register an explicit credential yourself in a similar way:
+#' ```
+#' my_cred <- git2r::cred_user_pass(
+#'   username = "janedoe",
+#'   password = "super_secret_password"
```

Comment:
I think we need to be very careful with this advice because if you run this code in your current session your password will get saved to `.Rhistory`, which you might accidentally expose.

Reply:
I could modify the example to show `Sys.getenv()` or `getPass::getPass()` as the way of handling the password field.

Reply:
I think `getPass::getPass()` would be good

Reply:
OK. Used for the analogous content re: `cred_ssh_key()` and the passphrase.